### ISSUE TITLE
Regression in handling numpy integers when passed to a Range trait with floats

### ISSUE THREAD

**Issue by prabhuramachandran** [enthought/traits#391]
Here is a regression in master it appears:
```
from traits.api import *
class A(HasTraits):
    x = Range(-10.0, 10.0)

a = A()
import numpy as np
a.x = np.int(1)  # OK
a.x = np.int64(1) # Raises a trait error!
```
This error does not seem to occur with traits version 4.6.0.

**Comment by rkern** [comment]
It works fine on `master` for me.

**Comment by prabhuramachandran** [comment]
Sorry, made a mistake typing the example, please try with the one updated above.

**Comment by rkern** [comment]
Still works for me. MacOS, Python 2.7.

However, on Python 3.6.0 and both traits 4.6.0 and `master`, I get the error.

**Comment by mdickinson** [comment]
Related: #292.

**Comment by mdickinson** [comment]
For the record, here's the traceback (with Traits 4.6.0):
```
taniyama:Desktop mdickinson$ edm run -e testing -- python traits_bug.py 
Traceback (most recent call last):
  File "traits_bug.py", line 8, in <module>
    a.x = np.int64(1) # Raises a trait error!
  File "/Users/mdickinson/.edm/envs/testing/lib/python3.6/site-packages/traits/trait_handlers.py", line 172, in error
    value )
traits.trait_errors.TraitError: The 'x' trait of an A instance must be -10.0 <= a floating point number <= 10.0, but a value of 1 <class 'numpy.int64'> was specified.
```
As @rkern indicates, this isn't a regression in Traits (though I guess it could be seen as a regression with respect to Python 2 -> Python 3): it fails for me on Traits 4.6.0, and I suspect has never worked on Python 3 for any version of Traits. The `Range` validator is expecting an instance of either `int` or `float`: on Python 2 (and not Windows), `np.int64` is a subclass of `int`, so the assignment works. In Python 3, it isn't, hence the error.

What we should really be doing here is re-using the existing `Int` and `Float` validators: a floating-point `Range` should accept anything that has either a `__float__` or an `__index__` method, and an integer range should accept anything that has an `__index__` method.

**Comment by prabhuramachandran** [comment]
Thanks for the explanation!  Do you think this has a good chance of being addressed in the next release?

**Comment by mdickinson** [comment]
Closing. This was fixed in #579 and release in Traits 6.0
